app/scripts/metamask-controller.js

```diff
@@ -3512,28 +3638,29 @@ export default class MetamaskController extends EventEmitter {
     const keyring = await this.getKeyringForDevice(deviceName, hdPath);
 
     keyring.setAccountToUnlock(index);
-    const oldAccounts = await this.keyringController.getAccounts();
     const keyState = await this.keyringController.addNewAccount(keyring);
-    const newAccounts = await this.keyringController.getAccounts();
-    this.preferencesController.setAddresses(newAccounts);
-    newAccounts.forEach((address) => {
-      if (!oldAccounts.includes(address)) {
-        const label = this.getAccountLabel(
-          deviceName === HardwareDeviceNames.qr
-            ? keyring.getName()
-            : deviceName,
-          index,
-          hdPathDescription,
-        );
-        // Set the account label to Trezor 1 /  Ledger 1 / QR Hardware 1, etc
-        this.preferencesController.setAccountLabel(address, label);
-        // Select the account
-        this.preferencesController.setSelectedAddress(address);
-      }
-    });
+    await this.accountsController.updateAccounts();
+    const newlyAddedAccount = this.accountsController.getSelectedAccount();
+    console.log(
+      this.getAccountLabel(
+        deviceName === HardwareDeviceNames.qr ? keyring.getName() : deviceName,
+        index,
+        hdPathDescription,
+      ),
+      index,
+      hdPathDescription,
+    );
+    await this.accountsController.setAccountName(
```

Comment:
In the previous logic we would potentially iterate over multiple new accounts, is it still possible to happen?

Reply:
This would be done on the accounts controller side